_drafts/restless-restfull.md

```diff
@@ -0,0 +1,456 @@
+```
+
+And it introduces an example of creating a resource to model something
+that is actually a process (see the "Nouns versus Verbs" part).
+
+This does seem at first as a way to introduce hacks instead of
+good design (specially if you are feeling RESTy), but almost all
+good ideas can be used as crappy hacks.
+Also the anemic feeling on APIs start to dissipate when you
+open yourself to this idea, so I embraced it.
+
+Later I encountered the concept of [reification](https://en.wikipedia.org/wiki/Reification),
+I was doing it for some time since it is what we do when we are programming,
+but I didn't have a name for it and never stopped to think on it
+as a valid design goal on a system.
+
+For example, compilers use this idea to make it easier to be ported to
+multiple platforms. What would be the reification ? The intermediary code.
+It is treated and tested on the compiler as real code, it is a valid
+testable product of the compilation process, but it does not exist, there
+is no machine on the world that can run it, it is a pure design creation
+made to make the system easier to port and test.
+
+So it can be your passport to an horrible API or to wonderful abstractions
+and a more testable API (I first heard about this for testing on this
+[great podcast](http://www.se-radio.net/2010/09/episode-167-the-history-of-junit-and-the-future-of-testing-with-kent-beck/)
+where Kent Beck talks about testing).
+
+# But what about the methods ?
+
+OK I got all excited with the state transfer thing and totally forgot
+that not being uniform on how you use a set of operations (methods) can
+be really hurtful to a distributed system.
+
+On a system where there are IDL's you usually don't have this problem
+since each interface has its own set of operations that is documented,
+there is less restrictions (also more space to inconsistencies in naming
+and other things, tradeoffs, no free lunch, etc).
+
+So REST and its more known implementation on top of HTTP has a fixed
+set of methods with a specific semantic for each one.
+
+Before going deeper on the semantics of the methods, you must ask yourself,
+why it is so important to use the correct semantic of each method ?
+
+The first that always came to my mind is that consistency makes your API
+easier to understand, you can use intuition to infer how to do stuff instead
+of always reading docs. Being able to use intuition to do something on an API
+is a sign of a great API to me. But when we are thinking about distributed
+systems there are other important things to consider.
+
+The most important one, that always bites us in the ass, is side effects.
+Our inability to properly handle side effects is what makes functional
+programming so attractive, we suck at it.
+
+So how does HTTP specifies state changes ? Lets take a look on the
+[spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html):
+
+```
+In particular, the convention has been established that the GET and HEAD
```

Comment:
The problem is that lots of APIs that work on collections require 'retrieval actions with no side effects' that GET nor HEAD can handle, and then people circumvent this with lots of ugly hacks with other methods. A great example is a search API, as it is a retrieval/filter operation in a collection but commonly requires a query language that do not fit in the URL params.

Reply:
I dont have a lot of experience with that, but I think I got your point. The search operation is side effect free but it has no clear way to be expressed using GET. I was checking the specs of the protocol:

https://tools.ietf.org/html/rfc2616#section-9

It does not seem like GET with a body would be wrong...perhaps it is another case of misguided guidelines and industry practice =P